File: frontend/lib/models/wallet.dart
/// Model Wallet
class Wallet {
  final double saldoPunti;
  final double? puntiEmessi;
  final double? puntiIncassati;
  final double? saldo;
  final DateTime? ultimoAggiornamento;

  Wallet({
    required this.saldoPunti,
    this.puntiEmessi,
    this.puntiIncassati,
    this.saldo,
    this.ultimoAggiornamento,
  });

  /// Factory da JSON
  factory Wallet.fromJson(Map<String, dynamic> json) {
    return Wallet(
      saldoPunti: _parseDouble(json['saldo_punti']),
      puntiEmessi: json['punti_emessi'] != null 
          ? _parseDouble(json['punti_emessi']) 
          : null,
      puntiIncassati: json['punti_incassati'] != null 
          ? _parseDouble(json['punti_incassati']) 
          : null,
      saldo: json['saldo'] != null 
          ? _parseDouble(json['saldo']) 
          : null,
      ultimoAggiornamento: json['ultimo_aggiornamento'] != null
          ? DateTime.parse(json['ultimo_aggiornamento'])
          : null,
    );
  }

  /// Helper per parsing sicuro double
  static double _parseDouble(dynamic value) {
    if (value is double) return value;
    if (value is int) return value.toDouble();
    if (value is String) return double.tryParse(value) ?? 0.0;
    return 0.0;
  }

  /// Formatta punti con 2 decimali
  String get saldoFormattato => saldoPunti.toStringAsFixed(2);

  /// Formatta saldo esercente
  String? get saldoEsercenteFormattato => 
      saldo != null ? saldo!.toStringAsFixed(2) : null;

  /// Verifica se ha saldo sufficiente
  bool hasSaldoSufficiente(double importo) => saldoPunti >= importo;

  /// Copy with
  Wallet copyWith({
    double? saldoPunti,
    double? puntiEmessi,
    double? puntiIncassati,
    double? saldo,
    DateTime? ultimoAggiornamento,
  }) {
    return Wallet(
      saldoPunti: saldoPunti ?? this.saldoPunti,
      puntiEmessi: puntiEmessi ?? this.puntiEmessi,
      puntiIncassati: puntiIncassati ?? this.puntiIncassati,
      saldo: saldo ?? this.saldo,
      ultimoAggiornamento: ultimoAggiornamento ?? this.ultimoAggiornamento,
    );
  }
}

File: backend/app/Models/Wallet.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;

/**
 * Model Wallet - Portafoglio digitale punti
 * 
 * Per clienti: solo saldo_punti è rilevante
 * Per esercenti: saldo_punti + punti_emessi + punti_incassati
 */
class Wallet extends Model
{
    use HasFactory;

    /**
     * Attributi assegnabili in massa
     *
     * @var array<int, string>
     */
    protected $fillable = [
        'user_id',
        'saldo_punti',
        'punti_emessi',
        'punti_incassati',
        'ultimo_aggiornamento'
    ];

    /**
     * Attributi da castare a tipi nativi
     *
     * @var array<string, string>
     */
    protected $casts = [
        'saldo_punti' => 'decimal:2',
        'punti_emessi' => 'decimal:2',
        'punti_incassati' => 'decimal:2',
        'ultimo_aggiornamento' => 'datetime'
    ];

    /**
     * Relazione: Utente proprietario del wallet
     */
    public function user()
    {
        return $this->belongsTo(User::class);
    }

    /**
     * Aggiungi punti al wallet
     * 
     * @param float $punti
     * @param string $motivo
     * @return bool
     */
    public function aggiungiPunti(float $punti, string $motivo = 'Aggiunta punti'): bool
    {
        $this->saldo_punti += $punti;
        $this->ultimo_aggiornamento = now();
        
        // Log::info("Wallet {$this->id}: +{$punti} punti. Nuovo saldo: {$this->saldo_punti}. Motivo: {$motivo}");
        
        return $this->save();
    }

    /**
     * Rimuovi punti dal wallet
     * 
     * @param float $punti
     * @param string $motivo
     * @return bool
     * @throws \Exception Se saldo insufficiente
     */
    public function rimuoviPunti(float $punti, string $motivo = 'Rimozione punti'): bool
    {
        if ($this->saldo_punti < $punti) {
            throw new \Exception("Saldo insufficiente. Disponibile: {$this->saldo_punti}, Richiesto: {$punti}");
        }

        $this->saldo_punti -= $punti;
        $this->ultimo_aggiornamento = now();
        
        // Log::info("Wallet {$this->id}: -{$punti} punti. Nuovo saldo: {$this->saldo_punti}. Motivo: {$motivo}");
        
        return $this->save();
    }

    /**
     * Registra punti emessi (solo per esercenti)
     * 
     * @param float $punti
     * @return bool
     */
    public function registraPuntiEmessi(float $punti): bool
    {
        $this->punti_emessi += $punti;
        $this->ultimo_aggiornamento = now();
        
        // Log::info("Wallet {$this->id} (Esercente): Emessi +{$punti}. Totale emessi: {$this->punti_emessi}");
        
        return $this->save();
    }

    /**
     * Registra punti incassati (solo per esercenti)
     * 
     * @param float $punti
     * @return bool
     */
    public function registraPuntiIncassati(float $punti): bool
    {
        $this->punti_incassati += $punti;
        $this->ultimo_aggiornamento = now();
        
        // Log::info("Wallet {$this->id} (Esercente): Incassati +{$punti}. Totale incassati: {$this->punti_incassati}");
        
        return $this->save();
    }

    /**
     * Calcola saldo esercente
     * Saldo = Punti Incassati - Punti Emessi
     * 
     * @return float
     */
    public function getSaldoEsercenteAttribute(): float
    {
        return (float) ($this->punti_incassati - $this->punti_emessi);
    }

    /**
     * Verifica se wallet ha saldo sufficiente
     * 
     * @param float $importo
     * @return bool
     */
    public function hasSaldoSufficiente(float $importo): bool
    {
        return $this->saldo_punti >= $importo;
    }

    /**
     * Scope: Filtra wallet con saldo positivo
     */
    public function scopeConSaldoPositivo($query)
    {
        return $query->where('saldo_punti', '>', 0);
    }

    /**
     * Scope: Filtra wallet di esercenti
     */
    public function scopeEsercenti($query)
    {
        return $query->whereHas('user', function($q) {
            $q->where('ruolo', 'esercente');
        });
    }

    /**
     * Scope: Filtra wallet di clienti
     */
    public function scopeClienti($query)
    {
        return $query->whereHas('user', function($q) {
            $q->where('ruolo', 'cliente');
        });
    }
}

